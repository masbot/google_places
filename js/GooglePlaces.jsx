import React, { Component } from 'react';
import axios from 'axios';
import Map from './Map';
import Search from './Search';
import List from './List';
import Spinner from './Spinner';
import preload from '../cities.json';

class GooglePlaces extends Component {
    state = {
        markers: [],
        searchTerm: '',
        map: null,
        zoom: 15,
        loading: false,
        city: { name: '', location: { lat: 0, lng: 0 } },
    };

    onSearchTermChange = event => {
        this.setState({ searchTerm: event.target.value });
    };

    onEnter = event => {
        if (event.key === 'Enter') {
            this.fetchMarker();
        }
    };

    onMapLoaded = map => {
        if (this.state.map != null) {
            return;
        }

        this.setState({
            map: map
        });
    };

    onMapMoved = () => {
        let location = JSON.stringify(this.state.map.getCenter());
        const lat = location.lat;
        const lng = location.lng;
        this.setState({city: {name: this.state.city.name, location: { lat: lat, lng: lng }}})

        if (this.state.searchTerm.length > 0) {
            this.fetchMarker();
        }
    };

    onClickDetail = targetMarker => {
        this.onClickMarker(targetMarker);
    };

    onClickMarker = targetMarker => {
        location.hash = '#' + targetMarker.id;
        this.setState({
            markers: this.state.markers.map(marker => {
                if (marker.id === targetMarker.id) {
                    marker.showInfo = true;
                } else {
                    marker.showInfo = false;
                }
                return marker;
            })
        });
    };

    onCloseMarker = targetMarker => {
        this.setState({
            markers: this.state.markers.map(marker => {
                if (marker.id === targetMarker.id) marker.showInfo = false;
                return marker;
            })
        });
    };

    onChooseCity = targetCity => {
        this.setState({ city: targetCity });
    };

    onClickReset = () => {
        console.log('reset')
        this.setState({ city: { name: '', location: { lat: 0, lng: 0 } }, searchTerm: '', markers: []});
    };

    fetchMarker = () => {
      this.setState({ loading: true });
      let location = JSON.stringify(this.state.map.getCenter());
      location = JSON.parse(location);
      const lat = location.lat;
      const lng = location.lng;
      const url = `https://maps.googleapis.com/maps/api/place/nearbysearch/json?location=${lat},${lng}&radius=500&&keyword=${this.state.searchTerm}&key=AIzaSyAB9Og-vXpoXkt8P9S_ZSWpFeRHMKHF74k`;
      console.log(url);
      axios({
        method: 'GET',
        url: url,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Headers': 'Origin, X-Requested-With, Content-Type, Accept',
          'Access-Control-Allow-Methods': 'GET',
          'Content-Type': 'application/x-www-form-urlencoded'
        }
      }).then(response => {
        console.log(response)
        this.setState({ markers: response.data.results, loading: false });
      });
    };

    render() {
        return (
            <div className="container-fluid">
                <Search
                    onEnter={this.onEnter}
                    onSearchTermChange={this.onSearchTermChange}
                    searchTerm={this.state.searchTerm}
                    city={this.state.city}
                />
                <div className="row">
                    <div className="col-md-4" style={{ width: '100%', height: '900px' }}>
                        <Map
                            map={this.state.map}
                            zoom={this.state.zoom}
                            markers={this.state.markers}
                            center={{ lat: this.state.city.location.lat, lng: this.state.city.location.lng }}
                            containerElement={<div style={{ height: '100%' }} />}
                            mapElement={<div style={{ height: '100%' }} />}
                            onMapLoaded={this.onMapLoaded}
                            onMapMoved={this.onMapMoved}
                            onClickMarker={this.onClickMarker}
                            onCloseMarker={this.onCloseMarker}
                        />
                    </div>
                    <div className="col-md-8" style={{ width: '100%', height: '900px', overflowX: 'scroll' }}>
                        {this.state.loading && <Spinner />}
                        <List
                            markers={this.state.markers}
                            onClickReset={this.onClickReset}
                            onChooseCity={this.onChooseCity}
                            onClickDetail={this.onClickDetail}
                            cities={preload.cities}
                            city={this.state.city}
                        />
                    </div>
                </div>
            </div>
        );
    }
}

export default GooglePlaces;
