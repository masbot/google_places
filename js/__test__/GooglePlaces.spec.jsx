import React from 'react';
import { shallow } from 'enzyme';
import GooglePlaces from '../GooglePlaces';

test('Search renders correctly', () => {
    const component = shallow(<GooglePlaces />);
    expect(component).toMatchSnapshot();
});
