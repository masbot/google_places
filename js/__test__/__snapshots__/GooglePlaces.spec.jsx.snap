// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<div
  className="container-fluid"
>
  <Search
    city={
      Object {
        "location": Object {
          "lat": 0,
          "lng": 0,
        },
        "name": "",
      }
    }
    onEnter={[Function]}
    onSearchTermChange={[Function]}
    searchTerm=""
  />
  <div
    className="row"
  >
    <div
      className="col-md-4"
      style={
        Object {
          "height": "900px",
          "width": "100%",
        }
      }
    >
      <withGoogleMap(Map)
        center={
          Object {
            "lat": 0,
            "lng": 0,
          }
        }
        containerElement={
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        }
        map={null}
        mapElement={
          <div
            style={
              Object {
                "height": "100%",
              }
            }
          />
        }
        markers={Array []}
        onClickMarker={[Function]}
        onCloseMarker={[Function]}
        onMapLoaded={[Function]}
        onMapMoved={[Function]}
        zoom={15}
      />
    </div>
    <div
      className="col-md-8"
      style={
        Object {
          "height": "900px",
          "overflowX": "scroll",
          "width": "100%",
        }
      }
    >
      <List
        cities={
          Array [
            Object {
              "id": 1,
              "location": Object {
                "lat": 40.728199,
                "lng": -73.9894738,
              },
              "name": "New York City",
              "state": "NY",
            },
            Object {
              "id": 2,
              "location": Object {
                "lat": 37.7653358,
                "lng": -122.4386605,
              },
              "name": "San Francisco",
              "state": "CA",
            },
            Object {
              "id": 3,
              "location": Object {
                "lat": 41.8333925,
                "lng": -88.0121478,
              },
              "name": "Chicago",
              "state": "IL",
            },
            Object {
              "id": 4,
              "location": Object {
                "lat": 29.816881,
                "lng": -95.6821568,
              },
              "name": "Houston",
              "state": "TX",
            },
            Object {
              "id": 5,
              "location": Object {
                "lat": 30.3409497,
                "lng": -81.6769833,
              },
              "name": "Jacksonville",
              "state": "FL",
            },
            Object {
              "id": 6,
              "location": Object {
                "lat": 35.4823226,
                "lng": -97.7600609,
              },
              "name": "Oklahoma City",
              "state": "OK",
            },
            Object {
              "id": 7,
              "location": Object {
                "lat": 34.0603946,
                "lng": -118.3302357,
              },
              "name": "Los Angeles",
              "state": "CA",
            },
            Object {
              "id": 8,
              "location": Object {
                "lat": 33.6050976,
                "lng": -112.4059151,
              },
              "name": "Phoenix",
              "state": "AZ",
            },
          ]
        }
        city={
          Object {
            "location": Object {
              "lat": 0,
              "lng": 0,
            },
            "name": "",
          }
        }
        markers={Array []}
        onChooseCity={[Function]}
        onClickDetail={[Function]}
        onClickReset={[Function]}
      />
    </div>
  </div>
</div>
`;
